File: 프로그래머스/0/181916. 주사위 게임 3/주사위 게임 3.cpp
#include <string>
#include <vector>
#include <unordered_set>
#include <algorithm>

using namespace std;

int solution(int a, int b, int c, int d) {
    int answer = 0;
    //중복 상관 X
    vector<int> x = {a,b,c,d};
    unordered_set<int> seen;
    
    vector<int> result;
    
    for (int t : x) {
        if (seen.insert(t).second) {
            result.push_back(t);
        }
    }
    
    //중복 상관 O
    vector<int>v = x;
    sort(v.begin(),v.end());
    v.erase(unique(v.begin(),v.end()),v.end());
    
    if (v.size()==1) return 1111 * v[0];
    else if (v.size()==2){
        if(count(x.begin(),x.end(),v[0]) == 2) return (v[0] + v[1]) * abs(v[0] - v[1]);
        else if (count(x.begin(),x.end(),v[0]) == 3) return (10*v[0]+v[1]) * (10*v[0]+v[1]);
        else {return (10*v[1]+v[0]) * (10*v[1]+v[0]);}
    }
    else if (v.size() == 3){
        for (int i = 0; i < 3; i++) {
            if (count(x.begin(), x.end(), v[i]) == 2) {
                int result = 1;

                for (int j = 0; j < 3; j++) {
                    if (i != j) {
                        result *= v[j];
                    }
                }

                return result;
            }
        }
    }
    else{
        return v[0];
    }
}

File: 프로그래머스/0/181916. 주사위 게임 3/주사위 게임 3.py
def solution(a, b, c, d):
    answer = 0
    vv = [a, b, c, d]
    num = list(set(vv))
    
    if len(num) == 1:
        return num[0] * 1111
    elif len(num) == 2:
        x, y = num
        if vv.count(x) == 2:
            return (x+y) * abs(x-y)
        elif vv.count(x) == 3:
            return (10*x + y) * (10*x + y)
        else:
            return (10*y + x) * (10*y + x)
            
    elif len(num) == 3:
        remain = [x for x in vv if vv.count(x)==1]
        return remain[0] * remain[1]
    else:
        return min(num)